ts/client/txPool.ts

```diff
@@ -1,40 +1,159 @@
+
+    /**
+     * Adds a transaction to the pool.
+     *
+     * @param tx Transaction to add to pool.
+     */
+    public async push(tx: Item): Promise<void> {
+        if (this.txCount >= this.maxSize) {
+            throw new PoolFullError(this.maxSize);
         }
+
+        const fromState = await this.stateStorage.get(tx.fromIndex.toNumber());
+        const tokenIDStr = fromState.tokenID.toString();
+        const tokenQueue = this.tokenIDStrToQueue[tokenIDStr];
+        if (!tokenQueue) {
+            throw new TokenNotConfiguredError(tokenIDStr);
+        }
+
+        tokenQueue.push(tx);
+        this.txCount++;
     }
-    pop() {
-        const tx = this.queue.pop();
-        if (!tx) throw new Error("Pool empty");
+
+    /**
+     * Removes the highest fee transaction for a token from the pool.
+     *
+     * @param tokenID Token to remove a transaction for.
+     * @returns Transaction from the pool
+     */
+    public pop(tokenID: BigNumber): Item {
+        const tokenIDStr = tokenID.toString();
+        const tokenQueue = this.tokenIDStrToQueue[tokenIDStr];
+        if (!tokenQueue) {
+            throw new TokenNotConfiguredError(tokenIDStr);
+        }
+        tokenQueue.sort(sortByFee);
+
+        const tx = tokenQueue.pop();
+        if (!tx) {
+            throw new TokenPoolEmpty(tokenID.toString());
+        }
+        this.txCount--;
         return tx;
     }
-    getValue(topN?: number) {
-        const size = topN ?? this.queue.length;
-        return sum(this.queue.slice(-size).map(tx => tx.fee));
+
+    /**
+     * Gets the highest value token transactions to process.
+     *
+     * Currently doesn't account for the token's exchange rate and
+     * prioritizes by summed fees.
+     *
+     * https://github.com/thehubbleproject/hubble-contracts/issues/572
+     * will change this behavior.
+     *
+     * @returns Object with the highest value tokenID, its feeReceiverID,
+     * and the sum of its total fees.
+     */
+    public async getHighestValueToken(): Promise<{
+        tokenID: BigNumber;
+        feeReceiverID: BigNumber;
+        sumFees: BigNumber;
+    }> {
+        if (!this.txCount) {
+            throw new PoolEmptyError();
+        }
+
+        let highValue = BigNumber.from(0);
+        let tokenID = BigNumber.from(-1);
+        for (const tokenIDStr of Object.keys(this.tokenIDStrToQueue)) {
+            const value = this.getQueueValue(tokenIDStr);
+            if (value.gt(highValue)) {
+                highValue = value;
+                tokenID = BigNumber.from(tokenIDStr);
+            }
+        }
+
+        const feeReceiverID = this.tokenIDStrToFeeRecieverID[
+            tokenID.toString()
+        ];
+        return {
+            tokenID,
+            feeReceiverID,
+            sumFees: highValue
+        };
+    }
+
+    private getQueueValue(tokenIDStr: string): BigNumber {
+        return sum(this.tokenIDStrToQueue[tokenIDStr].map(tx => tx.fee));
```

Comment:
Note that we have a queue size of 1024.
We have a commitment size of 32.
We pack 1 commitment at a time. 
So if we have a token A queue of 32 txs each worth $10 and a token B queue of 1024 tx each worth $1. The `getQueueValue ` considers token B queue is more valuable but token A queue is more profitable in a single pack. 

Not a big problem to fix now though.

Reply:
Ah good catch, that is something I can easily fix.

Reply:
Move to new issue

Reply:
https://github.com/thehubbleproject/hubble-contracts/issues/628